src/frontends/lean/structure_cmd.cpp

```diff
@@ -1289,6 +1287,7 @@ struct structure_cmd_fn {
             declare_no_confusion();
             declare_injective_lemmas();
         }
+        m_env = m_meta_info.m_attrs.apply(m_env, m_p.ios(), m_name);
         return m_env;
```

Comment:
Is this also a bug fix?

Reply:
Yes, or at least it is bringing `structure`'s behavior more in line with the other commands. Not sure where to document that behavior.

In my specific case, `@[derive has_reflect] structure` was trying to invoke `induction` before the structure was registered as an inductive.

Reply:
Before `m_env = m_meta_info.m_attrs.apply(m_env, m_p.ios(), m_name);`, we can add
```cpp
/*
Remark: it is important to apply attributes after the structure has been registered as an inductive 
datatype. For example, suppose we have `@[derive has_reflect] structure`, if the attributes were
applied earlier, the application would fail.
*/
```

Reply:
The need for replicating such a comment across all commands (and for the two fixes) probably points to an underlying design issue. I think we can do better and enforce the convention by moving attribute application into `parse_command`. Commands return a list of declaration names they want attributes applied to (multiple ones for mutual definitions, none for non-declaration commands). We can probably factor out even more code that way like applying docstrings, registering namespaces, etc.

Not only should this streamline the C++ code, it would also help anyone implementing a user command. What do you think?

Reply:
I don't like the idea of returning a "package of actions" to be executed in the parser.
This kind of "package" can get arbitrarily complicated: registering attributes, doc strings, etc.
It will get more complicated as we add/change features. Moreover, it is not clear whether the order we execute actions will always be the same for all commands. 
I think we can cleanup the code by providing a default "finalization procedure". This procedure can be built using smaller ones. If the default is not good for a specific command, then we assemble a new one using (some of) the smaller ones.

I'm also trying to minimize the amount of work on the parser since we have issue #1674. 
To be honest, I don't think it is feasible to attack issue #1674 this year. It is a major change.
First, redoing `parser`+`pp`+`elaborator` would be a major amount of work. Whenever, I fix bugs on the elaborator, it reminds me on how much stuff would have to be re-implemented. 
I think a more realistic path is to first write the new `parser` and a converter from `syntax` object to `pexpr`. In this way, we don't need to re-implement `pp` and the `elaborator`. If we succeed, then we try to implement the pretty printer. Then, we try the `elaborator`. 
Just a new `parser`, `syntax` object and `pp` would give us most of the benefits. At least, the ones that are visible to most users. These features are sufficient for implementing refactoring tools,  hole commands, and tactics that suggest replacements (e.g., sledgehammer-like tactics that run for awhile but produce a new tactic script that can be replayed efficiently). A new elaborator would be mainly useful for us and very sophisticated users.

Reply:
> I don't like the idea of returning a "package of actions" to be executed in the parser.
This kind of "package" can get arbitrarily complicated: registering attributes, doc strings, etc.

Note that we already have such a package in form of the `cmd_meta` structure. Intuitively, it would make sense to me if this information that is parsed independent of the command would also be exactly the one we could apply independent of the command.

> I think we can cleanup the code by providing a default "finalization procedure".

Okay, that would also work. A small advantage of the cmd_meta-returning approach is that we could check centrally that a non-declaration command indeed doesn't take attributes or a doc string, which would have found the `@[simp] axiom` bug. Alternatively, we could pass the cmd_meta by non-const reference and check in the end whether all information has been "consumed".

> I think a more realistic path is to first write the new parser and a converter from syntax object to pexpr. In this way, we don't need to re-implement pp and the elaborator. If we succeed, then we try to implement the pretty printer. Then, we try the elaborator.

Yes, I believe that is the only realistic approach. I hope I can soon estimate how much time I will be able to put into that project.